menu/cbs/menu_cbs_sublabel.c

```diff
@@ -807,6 +810,104 @@ static int action_bind_sublabel_netplay_room(
 }
 #endif
 
+static int action_bind_sublabel_playlist_entry(
+      file_list_t *list,
+      unsigned type, unsigned i,
+      const char *label, const char *path,
+      char *s, size_t len)
+{
+   settings_t *settings = config_get_ptr();
+   playlist_t *playlist = NULL;
+   const char *playlist_path = NULL;
+   const char *core_path = NULL;
+   const char *core_name = NULL;
+   
+   if (!settings->bools.playlist_show_sublabels)
+      return 0;
+   
+   /* Get current playlist */
+   playlist = playlist_get_cached();
+   if (!playlist)
+      return 0;
+   if (i >= playlist_get_size(playlist))
+      return 0;
+
+   /* Read playlist entry */
+   playlist_get_index(playlist, i, &playlist_path, NULL, &core_path, &core_name, NULL, NULL);
+   
+   /* Only add sublabel if a core is currently assigned */
+   if (string_is_empty(core_name) || string_is_equal(core_name, file_path_str(FILE_PATH_DETECT)))
+      return 0;
+   
+   /* Add core name */
+   snprintf(s, len, "%s %s",
+      msg_hash_to_str(MENU_ENUM_LABEL_VALUE_PLAYLIST_SUBLABEL_CORE),
+      core_name);
+   
+   /* Get runtime *if* 'content_runtime_log' is enabled
+    * NB: Runtime is currently stored in an independent
+    * 'content_runtime.lpl' file, similar to the content
+    * history. It therefore only really makes sense to
+    * check runtime when viewing the content history
+    * playlist. If runtime were added to all playlists
+    * (would be nice), we could do this trivially for all
+    * content. */
+   if (!settings->bools.content_runtime_log)
+      return 0;
+   
+   if (!string_is_equal(label, msg_hash_to_str(MENU_ENUM_LABEL_LOAD_CONTENT_HISTORY))
+         && !string_is_equal(label, msg_hash_to_str(MENU_ENUM_LABEL_HISTORY_TAB)))
+      return 0;
+   
+   /* Check that 'content_runtime.lpl' exists, and playlist has
+    * non-null path/core_path values... */
+   if (g_defaults.content_runtime && !string_is_empty(playlist_path) && !string_is_empty(core_path))
+   {
+      unsigned runtime_hours;
+      unsigned runtime_minutes;
+      unsigned runtime_seconds;
+      unsigned j;
+      
+      /* This is lame, but unless runtime is added to all playlists
+       * we can't really do it any other way...
+       * Search 'content_runtime.lpl' until we find the current
+       * content+core combo. */
+      for (j = 0; j < playlist_get_size(g_defaults.content_runtime); j++)
```

Comment:
I think the PR is great, the only concern I have is how much CPU overhead this would add. This will be executed for every sublabel per frame, so the bigger the playlist is, the more iterations there would be here.

Can you do some performance tests on lower-end hardware (like say 3DS or whatever) and see what kind of performance overhead we would be looking at here and if it would be a real concern?

Reply:
>  the only concern I have is how much CPU overhead this would add

That for loop doesn't sit well with me, either.

I was thinking about it last night - there are basically two ways to fix this 'properly':

- Add the playtime record to `content_history.lpl` itself. This may be best, but there would be some non-trivial complications to resolve...

- Search `content_runtime.lpl` only once each time the content history menu item is selected (probably using some kind of hash table). There would still be a performance impact, but the effects would be limited (i.e. in the worst case a slight pause when opening a very long history list, but no issues once it's open)

I'll have a look at this now. Give me a few hours and I'll report back.

> This will be executed for every sublabel per frame

I noticed this, but I'm not sure of the reasoning. It seems very wasteful. Surely items like UI strings should only be fetched when the menu changes...?

Reply:
Certainly, I think everything menu is queried, processed and rendered (that last one is unavoidable I guess) every frame
it would be desirable to populate and cache the data till you move (or once a second if sitting still) I guess.

Even rendering could be optimized so it only re-renders dirty regions instead of the whole thing (selected item, clock would be dirty regions, maybe the ticker in RGUI too)

Reply:
@fr500 Yes, I think there's quite a bit of room for optimisation.

RGUI is reasonably well behaved in that it only updates the buffer (querying menu entries, etc.) when the display has changed. But even with RGUI, `menu_entry_get()` is called twice each frame, here:

https://github.com/libretro/RetroArch/blob/44431ad227113bd94c4ba403e8ce95e80a40b0d7/menu/menu_input.c#L882

and here:

https://github.com/libretro/RetroArch/blob/44431ad227113bd94c4ba403e8ce95e80a40b0d7/menu/drivers/menu_generic.c#L231

which is wasteful because we're fetching several items (including sublabels) which aren't used...

Reply:
@twinaphex Okay, I've just pushed an update that changes the following:

- The content runtime is now only parsed when the content history playlist is loaded/selected via the menu. Generating the playlist sublabels is now a trivial thing.

- Bug fix: I discovered that the original runtime logging calculations had some unfortunate errors (almost typo kind of stuff...). This is now fixed.

- Bug fix: I also fixed an oversight where contentless cores (e.g. ScummVM) were not getting logged when shutting down via the 'Close Content' menu option.

I spent a while with the content history 'parsing'. Basically, for the size of playlist we're looking at (~100 entries for a typical history), a brute force linear search is faster than any kind of binary search or hashing tomfoolery. With a fully populated record of 100 entries, parsing the whole thing takes less time than I can measure on my crappy old workstation (Athlon II X3 435 - single thread rating of 1004 on PassMark), so I really don't think it's too bad (there is certainly no sort of lag with any of the menu drivers).

Do you think this is good enough?

I suppose we could improve things further by getting rid of the `Save content runtime log` option and always log runtime when history is enabled - this would ensure that the content and runtime playlists stay in sync, which should simplify any searching. However, I'm not sure that the performance gain is significant enough to worry about, and I kinda like having the option to turn runtime logging off (some people aren't going to want it, so is it right to force it on everyone?).

Anyway, please let me know what you think.

I still need to tidy up `menu_displaylist.c` to remove core names from entries when sublabels are turned on, but I'll make a separate PR for that since I don't want to confuse matters.

Reply:
Yes, that looks a lot better.

I agree with both you and @fr500 that we should really look into optimizing the menu in these areas you mention, particularly the idea with caching and avoiding recurrent function calls like menu_entry_get is all music to my ears. Ideally we can hose a lot of conditional logic out of the menu drivers themselves (so RGUI/XMB/MaterialUI/Ozone) so that they mainly have to concern themselves with presenting text/graphics that have already been processed/arranged elsewhere.

Anyway, I think this PR is good for merging, and you guys can separately research ways to optimize the menu in subsequent PRs.